build.gradle

```diff
@@ -1,23 +1,102 @@
+import java.util.Properties;
 
 plugins {
     id "java"
     // moditect for module-info.class injection and jlink
     id "org.moditect.gradleplugin" version "1.0.0-rc3"
     // jpackage plugin for packging jextract native app
     id "org.panteleyev.jpackageplugin" version "1.3.1"
+    // for downloading files
+    id "de.undercouch.download" version "5.0.2"
 }
 
 version = '1.0'
+project.ext.minimalJDKVersion = Runtime.Version.parse("18");
+project.ext.depsDir = "deps"
 
 def static checkPath(String p) {
-    if (!Files.exists(Path.of(p))) {
+    checkPath(Path.of(p));
+}
+
+def static checkPath(Path p) {
+    if (!Files.exists(p)) {
         throw new IllegalArgumentException("Error: the path ${p} does not exist");
     }
 }
 
+def static Runtime.Version getJDKVersion(String jdkHome) {
+    Path releaseFile = Path.of(jdkHome).resolve("release");
+    checkPath(releaseFile);
+
+    Properties props = new Properties();
+    try (InputStream input = Files.newInputStream(releaseFile)) {
+        props.load(input);
+    }
+
+    String foundVersionStr = props.getProperty("JAVA_VERSION");
+    if (foundVersionStr == null) {
+        throw new IllegalArgumentException("Could not determine java version from ${releaseFile}");
+    }
+
+    foundVersionStr = foundVersionStr.substring(1, foundVersionStr.length() - 1); // version in release file is quoted
+    Runtime.Version foundVersion = Runtime.Version.parse(foundVersionStr);
+
+    return foundVersion;
+}
+
+def static checkJDKRelease(String jdkHome, Runtime.Version expectedVersion) {
+    Runtime.Version foundVersion = getJDKVersion(jdkHome);
+    if (foundVersion.compareToIgnoreOptional(expectedVersion) < 0) {
+        throw new IllegalArgumentException("Incorrect build JDK version. " +
+            "Expected at least '${expectedVersion}', but found '${foundVersion}'");
+    }
+}
+
+def String downloadJDK() {
+    String jdkUrl = "https://download.java.net/java/GA/jdk18/43f95e8614114aeaa8e8a5fcf20a682d/36/GPL/openjdk-18_";
+    String archiveExt;
+    if (Os.isFamily(Os.FAMILY_WINDOWS)) {
+        jdkUrl += "windows-x64_bin.zip";
+        archiveExt = ".zip";
+    } else if (Os.isFamily(Os.FAMILY_MAC)) {
+        jdkUrl += "macos-x64_bin.tar.gz"
+        archiveExt = ".tar.gz";
+    } else {
+        jdkUrl += "linux-x64_bin.tar.gz"
+        archiveExt = ".tar.gz";
+    }
+
+    String jdkZipPath = ext.depsDir + "/jdk" + archiveExt;
+    download.run {
+        src jdkUrl
+        dest jdkZipPath
+        overwrite false // don't overwrite if it already exists
+    }
+
+    // extract
+    FileTree source = Os.isFamily(Os.FAMILY_WINDOWS) ? zipTree(jdkZipPath) : tarTree(jdkZipPath);
+    copy {
+        from source
+        into ext.depsDir
+    }
+
+    return ext.depsDir + "/jdk-18";
+}
+
+String jdkHome;
+if (project.hasProperty("jdk")) {
+    println("using JDK from 'jdk' project property");
+    jdkHome = project.property("jdk");
+    checkJDKRelease(jdkHome, minimalJDKVersion);
```

Comment:
we check minimum JDK. So.. jdk 19 will be accepted even after foreign-preview hits jdk 19. But that won't work with this version of jdk. I think it is better to accept only jdk 18 - as this jextract code/test/samples work only with jdk 18.